packages/xod-client/src/hinting/typeDeduction.js

```diff
@@ -0,0 +1,89 @@
+import * as R from 'ramda';
+import * as XP from 'xod-project';
+import { isAmong, foldMaybe, maybeProp, maybePath } from 'xod-func-tools';
+
+import * as PAT from '../project/actionTypes';
+
+import { getActingPatchPath } from './utils';
+
+// =============================================================================
+//
+// Type deduction
+//
+// =============================================================================
+
+// Run type deduction only for some actions
+// :: Project -> Action -> Boolean
+export const shallDeduceTypes = R.curry((project, action) =>
```

Comment:
I see a conceptual flaw with `shallDeduceTypes` implementation. It’s a whitelist instead of a blacklist and thus fragile. That is, it returns `false` (will not deduce types) by default and returns `true` only in some cases foreseen by a developer. It seems like a big source of bugs. I did not test it yet, but I’m quite sure from what I see, the following actions will not update deduced types:

- Node deletion
- Link deletion

They are just out of my head, who knows how many actions are actually forgotten.

Better if the function will return `true` (will deduce types) by default and `false` in some short-circuit cases. Please, emphasize it in comments as well.

Reply:
> but I’m quite sure from what I see, the following actions will not update deduced types

It will because of action type called "BULK_DELETE_ENTITIES" is listed.

I understand concerns about the whitelist, but I'm afraid that in this case almost all actions, except those listed, will be blacklisted. Is it OK?

Reply:
I think it is OK. In this case, the worst thing we get is a slow-down. In the current implementation, the worst thing is buggy behavior. I’d prefer the first. Being real, only actions from `./packages/xod-client/src/project/actionTypes.js` can lead to re-deduction (because only they change the `project` branch) and there are less than 50 actions there. So white-listing a few dozens of actions should not be a big deal: start with copy-paste and remove those which can affect the type deduction.